src/link/output.c

```diff
@@ -483,83 +471,109 @@ static uint16_t writeMapBank(struct SortedSections const *sectList,
-		return;
-
 	fputs("SUMMARY:\n", mapFile);
 
 	for (uint8_t i = 0; i < SECTTYPE_INVALID; i++) {
 		enum SectionType type = typeMap[i];
+		uint32_t nbBanks = sections[type].nbBanks;
 
 		// Do not output used space for VRAM or OAM
 		if (type == SECTTYPE_VRAM || type == SECTTYPE_OAM)
 			continue;
 
 		// Do not output unused section types
-		if (sections[type].nbBanks == 0)
+		if (nbBanks == 0)
 			continue;
 
+		uint32_t usedTotal = 0;
+
+		for (uint32_t bank = 0; bank < nbBanks; bank++) {
+			uint16_t used = 0;
+			struct SortedSections const *sectList = &sections[type].banks[bank];
+			struct SortedSection const *section = sectList->sections;
+			struct SortedSection const *zeroLenSection = sectList->zeroLenSections;
+
+			while (section || zeroLenSection) {
+				struct SortedSection const **pickedSection =
+					nextSection(&section, &zeroLenSection);
+
+				used += (*pickedSection)->section->size;
+				*pickedSection = (*pickedSection)->next;
+			}
+
+			usedTotal += used;
+		}
+
 		fprintf(mapFile, "\t%s: %" PRId32 " byte%s used / %" PRId32 " free",
-			sectionTypeInfo[type].name, usedMap[type], usedMap[type] == 1 ? "" : "s",
-			sections[type].nbBanks * sectionTypeInfo[type].size - usedMap[type]);
-		if (sectionTypeInfo[type].firstBank != sectionTypeInfo[type].lastBank ||
-		    sections[type].nbBanks > 1)
-			fprintf(mapFile, " in %d bank%s", sections[type].nbBanks,
-				sections[type].nbBanks == 1 ? "" : "s");
-		fputc('\n', mapFile);
+			sectionTypeInfo[type].name, usedTotal, usedTotal == 1 ? "" : "s",
+			nbBanks * sectionTypeInfo[type].size - usedTotal);
+		if (sectionTypeInfo[type].firstBank != sectionTypeInfo[type].lastBank
+		    || nbBanks > 1)
+			fprintf(mapFile, " in %d bank%s", nbBanks, nbBanks == 1 ? "" : "s");
+		putc('\n', mapFile);
 	}
 }
 
-// Writes the sym and/or map files, if applicable.
-static void writeSymAndMap(void)
+// Writes the sym file, if applicable.
+static void writeSym(void)
 {
-	if (!symFileName && !mapFileName)
+	if (!symFileName)
 		return;
 
-	uint32_t usedMap[SECTTYPE_INVALID] = {0};
-
 	symFile = openFile(symFileName, "w");
-	mapFile = openFile(mapFileName, "w");
+	if (!symFile)
+		return;
 
-	if (symFileName)
-		fputs("; File generated by rgblink\n", symFile);
+	fputs("; File generated by rgblink\n", symFile);
 
 	for (uint8_t i = 0; i < SECTTYPE_INVALID; i++) {
 		enum SectionType type = typeMap[i];
 
-		for (uint32_t bank = 0; bank < sections[type].nbBanks; bank++) {
-			struct SortedSections const *sect = &sections[type].banks[bank];
-
-			writeSymBank(sect, type, bank);
-			usedMap[type] += writeMapBank(sect, type, bank);
-		}
+		for (uint32_t bank = 0; bank < sections[type].nbBanks; bank++)
+			writeSymBank(&sections[type].banks[bank], type, bank);
 	}
 
-	writeMapSummary(usedMap);
-
 	closeFile(symFile);
+}
+
+// Writes the map file, if applicable.
+static void writeMap(void)
+{
+	if (!mapFileName)
+		return;
+
+	mapFile = openFile(mapFileName, "w");
+	if (!mapFile)
+		return;
```

Comment:
This (and the equivalent in the sym writing function) should report errors, instead of silently doing nothing. (You can test by passing a non-existent file, or `/dev/full`.)